### DISCUSSION TITLE
What is best practice for fetching data in Zustand?

### DISCUSSION THREAD

**Discussion by dante01yoon** [pmndrs/zustand#1415] [Q&A] [answered]
Hi, I'm new to Zustand. 
without using caching library like react-query or swr,
what would be best practice get and set zustand store for now?  🤔  

Is there more suitable way than initialize store in useEffect?

**Comment by dai-shi** [comment, marked as answer]
> What is best practice for fetching data in Zustand?

Async actions https://github.com/pmndrs/zustand#async-actions should be best practice?

**Reply by dante01yoon** [reply]
thank you for quick response! 

So, dispatch async actions in useEffect might be best solution before showing api responses form server? 

```js
const useFishStore = create((set) => ({
  fishies: {},
  fetch: async (pond, signal) => {

    const response = await fetch(pond, {signal})
    set({ fishies: await response.json() })
    
  },
}))

...
const { fetch: buyFish }  = useFishStore();   

useEffect(() => {
 const { signal, abort } = new AbortController();
 buyFish(myPond, signal);
 return () => {
   abort()
 }
},[])
```

**Reply by dai-shi** [reply]
I don't think `useEffect` is best practice. It's better to invoke actions in callbacks.

**Reply by dante01yoon** [reply]
can I ask simple example ?

**Reply by dai-shi** [reply]
For example, onClick handler:
```jsx
<button onClick={() => useFishStore.getState().fetch(...)}>Click me</button>
```

**Reply by dante01yoon** [reply]
hmm, but what if fetching server data should be happened without user interactive like button click?

**Reply by dai-shi** [reply]
Can it start fetching before rendering? The easiest is to start it at module load.

Anyway, `useEffect` is not that bad. Check out https://beta.reactjs.org/learn/you-might-not-need-an-effect for more info.

**Reply by dante01yoon** [reply]
I'll checkout what is module loading. thank you :)

**Reply by dai-shi** [reply]
by module loading, I meant like this:
```js
import create from 'zustand'

const useFishStore = create((set) => ({
  fishies: {},
  fetch: async (pond, signal) => {
    const response = await fetch(pond, {signal})
    set({ fishies: await response.json() })
  },
}))

// start fetching initially
useFishStore.getState().fetch(...)
```

**Reply by dante01yoon** [reply]
@dai-shi  seems get store outside of component and call it when component file used.

thanks!

**Reply by ewalti** [reply]
What's the data fetching story for async nuances like concurrent multi dispatches / race conditions in regards to this? Is it expected to handle this sort of thing in react-land via other hooks/solutions, then dispatch the result to the store? Or handle it in the store with some sort of query middleware?

**Reply by dai-shi** [reply]
Handle it in the store without middeware. Just start all fetching in a single action. Nothing goes wrong. If you would like take the advantage of Suspense, you could do it in the future: https://twitter.com/dai_shi/status/1580931674656817153

**Reply by Code-Victor** [reply]
is it possible to have `onLoading` and `onError` with Zustand async?

**Reply by dai-shi** [reply]
Why not. The library doesn't do anything, but you can.

**Reply by vaynevayne** [reply]
> @dai-shi seems get store outside of component and call it when component file used.

Wait a minute, this means that your interface is executed only once when it enters the application. Very often, the data is real-time. When we re-enter the page, we need to retrieve it. I don't think it's a problem to load the data for the first time in useEffect, because it means: page display-> fetch, which fits useEffect.

**Reply by dbritto-dev** [reply]
@vaynevayne let's use react query

**Reply by vaynevayne** [reply]
@dbritto-dev  Please don't, this is a client-side data request, this is not the problem that react-query wants to solve, using it forcibly, you will encounter a lot of trouble.

Before you use them, take a look at this article
https://juejin.cn/post/7223687532068077628
https://juejin.cn/post/7189518542780235836

**Reply by dbritto-dev** [reply]
@vaynevayne I didn't read it since isn't in english but that's the goal of having a data layer

**Reply by dante01yoon** [reply]
using react query is not a problem at all. it makes server data cache management in client side  handy.

**Comment by ItsTarik** [comment]
@vaynevayne its not OK pointing to chinese docs !

**Reply by vaynevayne** [reply]
Help others: use English as much as possible, if nothing, do not force
Ask for help: English forums must be in English
That's what I think.

To borrow eveny's words: 
> why should I sacrifice my time with my wife and children to save you time learning Chinese?


> @vaynevayne its not OK pointing to chinese docs !

**Reply by williammyuan** [reply]
小黑子,鸡脚露出来了吧

**Reply by Eireen** [reply]
@vaynevayne We do not require you to visit English-speaking programming communities. You are free to visit Chinese forums only.

**Comment by amoghkulkarnifr** [comment]
Hi all, I am also trying to find good resources about the same subject. I'm also new to Zustand and the best practices that one should follow for global state management. I was thinking of using SWR in addition to Zustand's persist middleware for getting the most out of it. 

What are some of the widely followed patterns for combining Zustand, persist and SWR?

**Reply by upromeer** [reply]
im also interested to know more about this....

**Reply by steaks** [reply]
I built [Leo Query](https://leoquery.com) to make integrating async data with Zustand easy. It handles caching, loading state, errors, stale data, etc. You may find it useful.

Here's a snippet from the [docs](https://leoquery.com/gettingStarted.html):

```tsx
import {create} from "zustand";
import {effect, query, hook} from "leo-query";

const useStore = create(() => ({
  increasePopulation: effect(increasePopulation),
  removeAllDogs: effect(removeAllDogs),
  dogs: query(fetchDogs, s => [s.increasePopulation, s.removeAllDogs]) // Re-fetch on populating increase or removal
}));

const useStoreAsync = hook(useStore, /*suspense*/false);

const YourComponent = () => {
  const dogs = useStoreAsync(s => s.dogs);
  
  if (dogs.isLoading) return <>Loading...</>;
  return <>{dogs.value}</>
};

```

**Comment by upromeer** [comment]
`store.tsx`
```tsx
const createDataStore = () =>
    createStore<TypeStore>()(
        immer(
            persist(
                (set) => ({
                    data: [],
                    index: 0,
                    saved: {},
                    setIndex: (index: number) => set({ index }),
                    seData: (data: TypeData[]) => set({ data }),
                    setSaved: (value: string) =>
                        set((state) => {
                            state.saved[state.data[state.index]] = value;
                        }),
                }),
                {
                    name: "data",
                }
            )
        )
    );
```

`provider.tsx`
```tsx
"use client";

import createDataStore from "./store";

const DataStoreContext = createContext<ReturnType<typeof createDataStore> | null>(null);

const DataStoreProvider = ({ children }: { children: ReactNode }) => {
    const storeRef = useRef<ReturnType<typeof createDataStore>>();
    
    if (!storeRef.current) {
        storeRef.current = createDataStore();
    }

    return (
        <DataStoreContext.Provider value={storeRef.current}>{children}</DataStoreContext.Provider>
    );
};


const useDataStore = <T,>(selector: (store: IDataStore) => T): T => {
    const dataStoreContext = useContext(DataStoreContext);

    if (!dataStoreContext) {
        throw new Error(`useDataStore must be use within DataStoreProvider`);
    }

    //// *useStoreWithEqualityFn for preventing re-rendring* ////
    return useStoreWithEqualityFn(dataStoreContext, selector);
};
```

` layout.tsx`
```tsx
import DataStoreProvider from "./provider";

export default function RootLayout({
    children,
}: Readonly<{
    children: ReactNode;
}>) {
    return <DataStoreProvider>{children}</DataStoreProvider>;
}
```

`page.tsx`
```tsx
"use client";

import useDataStore from "./provider";

export default function Home() {
  const setData = useDataStore((state) => state.setData);

  const { isLoading } = useQuery({
    queryKey: ["data"],
    queryFn: async () => {
      const response = await fetch("/data.json");
      return await response.json();
    },
    onSuccess(data) {
      setData(data);
    },
  });

  return <main>{isLoading ? <Supense /> : <Comp />}</main>;
}
```

**Reply by upromeer** [reply]
in `provider.tsx` there is `useStoreWithEqualitFn` for preventing re-render.

@dbritto-dev @dai-shi

is this correct implementation of zustand with react-query and nextjs?
or im doing something wrong?

**Comment by igortas** [comment]
@dai-shi 
I'm trying to achieve very similar scenario as  @upromeer, but I like to connect ```tanstack-query```, ```zustand``` and ```router``` loaders.
My idea is, just working with ```setState``` and ```zustand``` do the rest, with providing ```queryClient``` as param.

My current implementation:

Routes:

```
     <Route
          path='/'
          element={<Home />}
          loader={async () => {
            await fetchUser(queryClient);
            return null;
          }}
        >
```

userstore

```
export const useUserStore = create<UserState>((set, get) => ({
  user: {},
  fetchUser: async (queryClient: QueryClient) => {
    const user = await queryClient.ensureQueryData<User>({
      queryKey: ['user'],
      queryFn: () => fetcher
    });
    set({ user });
  }
}));

```

I don't see network request in ```devtools``` from the loader,  with just using the ```fetchUser``` setter and providing the ```queryClient``` instance.
Is it possible this scenario, or I need to set the ```fetch``` keyword exactly in ```zustand``` store, so will be fired on any ```set``` action?

**Reply by dai-shi** [reply]
I'm not super familiar with `router` nor `tanstack-query`, but the code looks good to me. How are you getting `queryClient` and `fetchUser` in `<Route>`?

> I need to set the fetch keyword exactly in zustand store, so will be fired on any set action?

Not sure if I get it.

**Comment by igortas** [comment]
> I'm not super familiar with `router` nor `tanstack-query`, but the code looks good to me. How are you getting `queryClient` and `fetchUser` in `<Route>`?
> @dai-shi I've made it work, was issue from the ``router``` side, I was using different type of router option that was ignoring loader functions....
> > I need to set the fetch keyword exactly in zustand store, so will be fired on any set action?
> I was thinking that the only way ```zustand``` will fetch some data, if u name the ```store``` function only as ```fetch```, not ```fetchUser```... But I was wrong...
> Not sure if I get it.

**Comment by leookun** [comment]
It's useful for me.
```jsx
const Com = () => {
  if (!state) {
    throw stateSetter();
  }
  return xxx;
};
```

```jsx
<ErrorBoundary FallbackComponent={...}>
  <React.Suspense fallback={...}>
    <Com />
  </React.Suspense>
</ErrorBoundary>
```
